llvm/lib/Transforms/Scalar/DFAJumpThreading.cpp

```diff
@@ -439,18 +449,31 @@ struct MainSwitch {
         // initial switch values that can be ignored (they will hit the
         // unthreaded switch) but this assumption will get checked later after
         // paths have been enumerated (in function getStateDefMap).
+
+        // If the unpredictable value comes from the same inner loop it is
+        // likely that it will also be on the enumerated paths, causing us to
+        // exit after we have enumerated all the paths. This heuristic save
+        // compile time because a search for all the paths can become expensive.
+        if (EarlyExitHeuristic && LI->getLoopFor(SI->getParent()) ==
+                                      LI->getLoopFor(CurrentIncomingBB)) {
```

Comment:
I don't really get why this cares about the loop of the incoming BB, rather than the loop where the instruction is defined.

Reply:
Also, what about the case where the incoming BB is a sub-loop? Wouldn't we want to bail out in that case as well?

Reply:
> Also, what about the case where the incoming BB is a sub-loop? Wouldn't we want to bail out in that case as well?

Makes sense, I have added the check.

Reply:
> I don't really get why this cares about the loop of the incoming BB, rather than the loop where the instruction is defined.

I don't have a good answer for this, except that the compile time won't will be any better if I check the instruction's loop. 

The added test3 shows this case. Even without this patch `getStateDefMap` rejects the switch because it checks for the incoming BB and not the value's defining BB,  I am not sure why the pass was written to disallow such cases. I think the switch is still threadable but maybe there are additional complications I am not aware of?

Reply:
> I don't really get why this cares about the loop of the incoming BB, rather than the loop where the instruction is defined.

Also, I notice that doing so loses a couple more transformations. I haven't looked into those though.